dlobron/ChugBot
# Schedule builder

### Description
Implements a mechanism to create custom schedules for every camper based on a set template. Allows for the creation, saving, and deletion of templates, as well as designating which template applies to which edah. Also provides a way to display the assignments from multiple blocks (overriding an assigned one by block, e.g. Chug Aleph for Block 1 instead of Chug Aleph for Block 2 when all other chugim shown are for Block 2).

In addition, this PR updates the entire site to use Bootstrap 5 (from Bootstrap 3).

### Specific Features
- Create a custom schedule which can be printed for each camper with their own chug assignments
- Preview the schedule before printing it
- Editor included in webpage to allow flexible formatting/editing
- Buttons enable the quick insertion of placeholders corresponding to the variables which change by camper (name, edah, chug assignments, etc)
- Overriding by block
- Save schedules for easy retrieval
    - Includes a clear way to both update a saved template and save a new template, depending on your desire
- Delete saved schedules (plus a clear warning)
- Schedules can be specifically assigned to certain edot, and be made not available for others
- Entire website overhauled to utilize Bootstrap 5 (upgraded from Bootstrap 3)

### Testing Instructions
In the directory where the repo is stored locally, run this command:
```
./start.sh
```
Then, navigate to [http://127.0.0.1:8000/](http://127.0.0.1:8000/) on that device and test normally. No coding should be required for basic testing.

### Installation/Upgrade Instructions
This update requires 2 new tables in the database:
1. Stores the schedules (id, name, schedule)
2. Stores the relation between edah and schedule to track which edot can see which schedules (schedule id, edah id)

To create these tables in the live applications of the application, sign into the SQL server, select the correct database, and run the following to create the structure which will be populated as the application runs:

```
--
-- Table structure for table `schedules`
--

DROP TABLE IF EXISTS `schedules`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!50503 SET character_set_client = utf8mb4 */;
CREATE TABLE `schedules` (
  `schedule_id` int NOT NULL AUTO_INCREMENT,
  `name` varchar(50) CHARACTER SET utf8mb3 COLLATE utf8mb3_unicode_ci NOT NULL,
  `schedule` TEXT CHARACTER SET utf8mb3 COLLATE utf8mb3_unicode_ci NOT NULL,
  PRIMARY KEY (`schedule_id`)
) ENGINE=InnoDB AUTO_INCREMENT=5 DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;

--
-- Table structure for table `edot_for_schedules`
--

DROP TABLE IF EXISTS `edot_for_schedule`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!50503 SET character_set_client = utf8mb4 */;
CREATE TABLE `edot_for_schedule` (
  `schedule_id` int NOT NULL,
  `edah_id` int NOT NULL,
  PRIMARY KEY (`schedule_id`,`edah_id`),
  KEY `fk_edah_id` (`edah_id`),
  CONSTRAINT `edot_for_schedule_ibfk_1` FOREIGN KEY (`schedule_id`) REFERENCES `schedules` (`schedule_id`) ON DELETE CASCADE ON UPDATE CASCADE,
  CONSTRAINT `edot_for_schedule_ibfk_2` FOREIGN KEY (`edah_id`) REFERENCES `edot` (`edah_id`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;
```
